### PR TITLE
Default repo rework

### PR DESCRIPTION
**Pull request by tejal29**
Reviving: #2356 ( unfortunately, that branch is protected and i can't push to it)
I made a bunch of changes to this branch hence created a new PR
Fixes: #1534

Re-worked GCR image substitutions as per this internal DD https://docs.google.com/document/d/1jL29npnNaz3lTz62gBjrIuT8q4wDqGqnqSV0_LwEk1o/edit#

List of changed
- [x] uses a config to activate rewrite substitution rules.
- [x] Add doc changes 

**Output***
With no `reconstructStrategy` set, 
```
 (default-repo-rework)../../out/skaffold build -d gcr.io/tejal-test
Generating tags...
 - gcr.io/k8s-skaffold/skaffold-example -> gcr.io/tejal-test/gcr.io/k8s-skaffold/skaffold-example:v1.9.0-50-g2903de0fc-dirty
Checking cache...
 - gcr.io/k8s-skaffold/skaffold-example: Found Remotely
```
Set the `reconstruct-strategy` to `true`
```
 (default-repo-rework)../../out/skaffold config set --global reconstruct-strategy true
set global value reconstruct-strategy to true
```
Rebuild
```
 (default-repo-rework)../../out/skaffold build -d gcr.io/tejal-test
Generating tags...
 - gcr.io/k8s-skaffold/skaffold-example -> gcr.io/tejal-test/skaffold-example:v1.9.0-50-g2903de0fc-dirty
Checking cache...
 - gcr.io/k8s-skaffold/skaffold-example: Found Remotely
 (default-repo-rework)
```

Finally set it to false
```
../../out/skaffold config set --global reconstruct-strategy false
set global value reconstruct-strategy to false
```
Rebuild
```
../../out/skaffold build -d gcr.io/tejal-test
Generating tags...
 - gcr.io/k8s-skaffold/skaffold-example -> gcr.io/tejal-test/gcr.io/k8s-skaffold/skaffold-example:v1.9.0-50-g2903de0fc-dirty
Checking cache...
 - gcr.io/k8s-skaffold/skaffold-example: Found Remotely
```

### CODE REVIEW

**Comment by tejal29** [comment]
> Haven't looked into this in any depth yet, but a few thoughts from a quick glance:

Thanks Brian. 
> 
> * This is a breaking change, and there may be people depending on the old-style of mapping. Is it worth having a flag to restore the old mapping?
This will change the `image` registry for GCR.  We can add a flag to restore old mapping. The reason i am not too inclined towards it is because fixes the way we used to handle GCR registry.  Folks should stop depending on old-style mapping.  But i can see the concern.

> * we should support our new Artifact Registry too

We could definitely do that in another PR. This PR add the framework and supports only GCR for now. 

> * It looks like you're saying that only GCR supports multiple path components? But IIRC Nexus does too. Should this be configurable? 
Hmm, that is good point. I could not find anything useful here 
Not really. Its the only one i implemented. In the DD, you will see there are others two which support multiple paths. We can definitely add a base `MultiPathRegistry` later as we add more. 
Right now, i feel what we have is enough for GCR. 

> Can we determine the registry type by querying the registry directly?
That is a good point. I could not find anything in the spec here https://docs.docker.com/registry/spec/api/

File: pkg/skaffold/docker/default_repo_test.go [hunk 1]
```diff
@@ -48,13 +48,25 @@ func TestImageReplaceDefaultRepo(t *testing.T) {
...
 			image:         "eu.gcr.io/project/registry",
 			defaultRepo:   "eu.gcr.io/project",
 			expectedImage: "eu.gcr.io/project/registry",
 		},
 		{
-			description:   "image has shared prefix with defaultRepo",
+			description:   "default repo registry is in another domain and different subpaths",
+			image:         "gcr.io/project/subpath/registry",
+			defaultRepo:   "eu.gcr.io/project/defaultRepoSubpath",
+			expectedImage: "eu.gcr.io/project/defaultRepoSubpath/subpath/registry",
```
Comment (tejal29) [inline pkg/skaffold/docker/default_repo_test.go:60]: Add `project2`

**Comment by tejal29** [comment]
@nkubala added docs for the default repository behavior like you mentioned.

File: pkg/skaffold/docker/default_repo.go [hunk 2]
```diff
@@ -49,6 +58,38 @@ func SubstituteDefaultRepoIntoImage(defaultRepo string, image string) (string, e
...
+	// replace registry in parsed name
+	var replaced string
+	reg, image := splitImage(parsed.BaseName)
+	defaultRegistry := registry.GetRegistry(defaultRepo)
+	newReg := reg.Update(defaultRegistry)
+	if newReg.Type() == reg.Type() {
+		replaced = newReg.Name() + "/" + image
+	} else {
+		replaced = newReg.Name() + "/" + escapeRegex.ReplaceAllString(parsed.BaseName, "_")
+	}
```
Comment (briandealwis) [inline pkg/skaffold/docker/default_repo.go:80]: I found this section really hard to figure out what was going on.  It all seems to hinges on Registry.Update() returning its argument (`defaultRegistry`) unless they're both `GCRRegistry`.

File: pkg/skaffold/docker/default_repo.go [hunk 3]
```diff
@@ -19,16 +19,25 @@ package docker
...
 const maxLength = 255
 
 var (
-	escapeRegex = regexp.MustCompile(`[/._:@]`)
 	prefixRegex = regexp.MustCompile(`(.*\.)?gcr.io/[a-zA-Z0-9-_]+/?`)
+	escapeRegex = regexp.MustCompile(`[/._:@]`)
 )
 
-func SubstituteDefaultRepoIntoImage(defaultRepo string, image string) (string, error) {
+func SubstituteDefaultRepoIntoImage(defaultRepo string, image string, rewriteStrategy bool) (string, error) {
```
Comment (briandealwis) [inline pkg/skaffold/docker/default_repo.go:33]: TBH I found this code hard to parse and reason through.  I had to scroll back and forth to track what was going on.

I wonder if it could be simplified by instead representing the strategies as objects, and the appropriate strategy is chosen based on the defaultRepo and rewriteStrategy.

File: pkg/skaffold/docker/default_repo.go [hunk 4]
```diff
@@ -49,6 +58,38 @@ func SubstituteDefaultRepoIntoImage(defaultRepo string, image string) (string, e
...
+	}
+
+	parsed, err := ParseReference(image)
+	if err != nil {
+		return "", err
+	}
+
+	// replace registry in parsed name
+	var replaced string
+	reg, image := splitImage(parsed.BaseName)
```
Comment (briandealwis) [inline pkg/skaffold/docker/default_repo.go:73]: `reg` is really `repo` (since _registry_ is the host, like gcr.io or docker.io)

File: pkg/skaffold/docker/default_repo.go [hunk 5]
```diff
@@ -49,6 +58,38 @@ func SubstituteDefaultRepoIntoImage(defaultRepo string, image string) (string, e
...
+	defaultRegistry := registry.GetRegistry(defaultRepo)
```
Comment (briandealwis) [inline pkg/skaffold/docker/default_repo.go:74]: s/GetRegistry/NewRegistry/`?

File: docs/content/en/docs/environment/image-registries.md [hunk 6]
```diff
@@ -49,29 +49,66 @@ Automated image name rewriting strategies are determined based on the default-re
      rewritten image:   aws_account_id.dkr.ecr.region.amazonaws.com/gcr_io_k8s-skaffold_skaffold-example1
     ```
 
-* default-repo begins with "gcr.io" (special case - as GCR allows for infinite deep image repo names)
-  * **strategy**: concat unless prefix matches
-  * **example1**: prefix doesn't match:
+* GCR Default repository.
+
+  For GCR default registry, Skaffold provides 2 strategies:
+  1. **Concatenation Strategy**: Concatenation strategy is the default strategy. In this strategy, the resulting image will have paths concatenated unless prefix matches.
```
[the remark this answers is not in the document]
Comment (tejal29) [reply docs/content/en/docs/environment/image-registries.md:70]: for now, we are going to make this default and slowly make the swap.

File: pkg/skaffold/config/util.go [hunk 7]
```diff
@@ -150,17 +150,23 @@ func getConfigForKubeContextWithGlobalDefaults(cfg *GlobalConfig, kubeContext st
...
+	if cliValue != nil {
+		return *cliValue, useReconstructStrategy(cfg), nil
+	}
+	return cfg.DefaultRepo, useReconstructStrategy(cfg), nil
+}
 
-	return cfg.DefaultRepo, nil
+func useReconstructStrategy(cfg *ContextConfig) bool {
+	if cfg.ReconstructStrategy == nil {
+		return false
```
[the remark this answers is not in the document]
Comment (tejal29) [reply pkg/skaffold/config/util.go:167]: We are not changing the default right now. When i demo-ed this feature in the fixit demo week, everyone agreed, we should not change the default behavior.

File: pkg/skaffold/docker/default_repo.go [hunk 8]
```diff
@@ -49,6 +58,38 @@ func SubstituteDefaultRepoIntoImage(defaultRepo string, image string) (string, e
 	return replaced, nil
 }
 
+func substituteDefaultRepoIntoImageNew(defaultRepo string, image string) (string, error) {
+	if defaultRepo == "" {
+		return image, nil
+	}
+
```
[the remark this answers is not in the document]
Comment (tejal29) [reply pkg/skaffold/docker/default_repo.go:65]: done

File: pkg/skaffold/docker/default_repo.go (hunk 2 above)
Comment (tejal29) [reply pkg/skaffold/docker/default_repo.go:80]: or both are `GenericRegistry`. I added some more test cases which will help clarify this more.

File: pkg/skaffold/docker/default_repo.go (hunk 3 above)
Comment (tejal29) [reply pkg/skaffold/docker/default_repo.go:33]: I refactored some more to keep the changes minimal. Please take a look and we can sync in person tomorrow?

File: pkg/skaffold/config/util.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (briandealwis) [reply pkg/skaffold/config/util.go:167]: Ok — though it came up in the IDE HaTS surveys as people didn't understand why they saw double gcr.io!

File: pkg/skaffold/docker/default_repo.go [hunk 9]
```diff
@@ -17,41 +17,40 @@ limitations under the License.
...
 
 const maxLength = 255
 
-var (
-	escapeRegex = regexp.MustCompile(`[/._:@]`)
-	prefixRegex = regexp.MustCompile(`(.*\.)?gcr.io/[a-zA-Z0-9-_]+/?`)
-)
-
-func SubstituteDefaultRepoIntoImage(defaultRepo string, image string) (string, error) {
+func SubstituteDefaultRepoIntoImage(defaultRepo string, image string, rewriteStrategy bool) (string, error) {
```
Comment (briandealwis) [inline pkg/skaffold/docker/default_repo.go:27]:
```
// SubstituteDefaultRepoIntoImage rehouses the provided image ref onto the defaultRepo.
// There are several approaches used to rehouse the image ref, as determined by the defaultRepo's
// registry type and the rewriteStragegy.
```

File: pkg/skaffold/docker/default_repo.go [hunk 10]
```diff
@@ -17,41 +17,40 @@ limitations under the License.
...
+	}
 	if parsed.Tag != "" {
 		replaced = replaced + ":" + parsed.Tag
 	}
 	if parsed.Digest != "" {
 		replaced = replaced + "@" + parsed.Digest
 	}
-
-	return replaced, nil
+	return truncate(replaced), nil
```
Comment (briandealwis) [inline pkg/skaffold/docker/default_repo.go:48]: The truncation should be on the image name, not the full ref.  That is, the tag + digest don't count.  

Hmm... re-reading the [OCI distribution spec](https://github.com/opencontainers/distribution-spec/blob/master/spec.md#overview) it's actually the repository name portion that must be less than 256 — the registry hostname doesn't count! 🙀

File: pkg/skaffold/docker/default_repo.go [hunk 11]
```diff
@@ -17,41 +17,40 @@ limitations under the License.
...
 		return "", err
 	}
 
-	replaced := replace(defaultRepo, parsed.BaseName)
+	var replaced string
+	if rewriteStrategy {
+		replaced = substituteDefaultRepoIntoImageWithRewrite(defaultRepo, parsed.BaseName)
+	} else {
+		replaced = replace(defaultRepo, parsed.BaseName)
+	}
```
Comment (briandealwis) [inline pkg/skaffold/docker/default_repo.go:41]: The truncation should happen here, before we add the tag and/or digest.
```suggestion
	}
	replaced = truncate(replaced)
```

File: pkg/skaffold/docker/default_repo.go [hunk 12]
```diff
@@ -60,11 +59,10 @@ func replace(defaultRepo string, baseImage string) string {
...
-		return truncate(defaultRepo + "/" + baseImage)
+		// prefixes don't match, concatenate image string
+		return defaultRepo + "/" + baseImage
 	}
-
-	return truncate(defaultRepo + "/" + escapeRegex.ReplaceAllString(baseImage, "_"))
+	return defaultRepo + "/" + registry.ESCRegex.ReplaceAllString(baseImage, registry.ReplaceStr)
 }
 
 func truncate(image string) string {
```
Comment (briandealwis) [inline pkg/skaffold/docker/default_repo.go:68]: Add a todo here that the OCI truncation rules apply to the repository name, not the full image name.

File: pkg/skaffold/docker/default_repo_test.go [hunk 13]
```diff
@@ -24,76 +24,130 @@ import (
...
+		{
+			description:      "aws",
+			image:            "gcr.io/some/registry",
+			defaultRepo:      "aws_account_id.dkr.ecr.region.amazonaws.com",
+			expectedImage:    "aws_account_id.dkr.ecr.region.amazonaws.com/gcr_io_some_registry",
+			expectedImageNew: "aws_account_id.dkr.ecr.region.amazonaws.com/gcr_io_some_registry",
+		},
+		{
+			description:      "aws over 255 chars",
+			image:            "gcr.io/herewehaveanincrediblylongregistryname/herewealsohaveanabnormallylongimagename/doubtyouveseenanimagethislong/butyouneverknowdoyouimeanpeopledosomecrazystuffoutthere/goodluckpushingthistoanyregistrymyfriend",
```
Comment (briandealwis) [inline pkg/skaffold/docker/default_repo_test.go:119]: Can you add an image tag to verify the truncation is only on the image name?

**Comment by tejal29** [comment]
This is on my back burner now.Closing this. Will re-open if i plan to work again,

**Comment by blaggacao** [comment]
Breadcrumb trail to use case of abstract reasoning: https://github.com/GoogleContainerTools/skaffold/issues/4858#issuecomment-703889238 &mdash; closing there, please reopen here.
